processing/src/main/java/org/apache/druid/segment/join/JoinVirtualColumnSplit.java

```diff
@@ -0,0 +1,148 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.join;
+
+import org.apache.druid.query.filter.Filter;
+import org.apache.druid.segment.VirtualColumn;
+import org.apache.druid.segment.VirtualColumns;
+import org.apache.druid.segment.column.RowSignature;
+import org.apache.druid.segment.join.filter.JoinFilterSplit;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+public class JoinVirtualColumnSplit
+{
+  /**
+   * Split {@link VirtualColumns} into those which can have inputs which exist on the base table and those which can
+   * be deferred until after the join due to not being present in the base table and thus all null
+   */
+  public static JoinVirtualColumnSplit split(
+      RowSignature baseTableSignature,
+      VirtualColumns virtualColumns,
+      JoinFilterSplit filterSplit
+  )
+  {
+    final Comparator<VirtualColumn> sorter = Comparator.comparing(VirtualColumn::getOutputName);
+    final SortedSet<VirtualColumn> preJoinVirtualColumns = new TreeSet<>(sorter);
+    final SortedSet<VirtualColumn> postJoinVirtualColumns = new TreeSet<>(sorter);
+    final Set<String> baseColumnNames = new HashSet<>(baseTableSignature.getColumnNames());
+    final Set<String> baseFilterColumns = filterSplit.getBaseTableFilter()
+                                                     .map(Filter::getRequiredColumns)
+                                                     .orElse(Collections.emptySet());
+    final Set<String> unresolvedNames = new HashSet<>();
+
+    final Set<VirtualColumn> unresolvedSet = new HashSet<>();
+    // do a first pass to eliminate those which depend directly on the base table
+    for (VirtualColumn vc : virtualColumns.getVirtualColumns()) {
+      if (baseColumnNames.containsAll(vc.requiredColumns()) || baseFilterColumns.contains(vc.getOutputName())) {
+        preJoinVirtualColumns.add(vc);
+        baseColumnNames.add(vc.getOutputName());
+        final Set<VirtualColumn> dependencies = resolveDependencies(virtualColumns, vc);
+        for (VirtualColumn dep : dependencies) {
+          preJoinVirtualColumns.add(dep);
+          postJoinVirtualColumns.remove(dep);
+          unresolvedSet.remove(dep);
+          baseColumnNames.add(dep.getOutputName());
+        }
```

Comment:
this would only have meaning for virtual columns that baseFilterColumn refers to, correct? If the baseColumnNames contain all required columns of vc, it would mean the dependencies are already resolved. can you confirm?

Reply:
yea, I could put it under an if that checks for that case, or separate it out into an else if you prefer